include/alpaka/block/shared/st/detail/BlockSharedMemStMemberImpl.hpp

```diff
@@ -0,0 +1,190 @@
+#endif
+
+            //! Give the pointer to an exiting variable
+            //!
+            //! @tparam T type of the variable
+            //! @param id unique id of the variable
+            //! @return nullptr if variable with id not exists
+            template<typename T>
+            T* getVar(uint32_t id) const
+            {
+                // Get first meta data header.
+                MetaData* metaDataPtr = reinterpret_cast<MetaData*>(m_mem);
+                // Offset in bytes to the next meta data header behind the variable.
+                uint32_t off = 0;
+
+                // Last meta data header offset is zero
+                while(metaDataPtr->offset != 0)
+                {
+                    off = metaDataPtr->offset;
+                    if(metaDataPtr->id == id)
+                        break;
+                    // next meta data header
+                    metaDataPtr = reinterpret_cast<MetaData*>(m_mem + off);
+                }
+
+                // Last header, variable with id not found.
+                if(metaDataPtr->offset == 0)
+                {
+                    return nullptr;
+                }
+
+                return reinterpret_cast<T*>(&m_mem[off - sizeof(T)]);
+            }
+
+        private:
+            //! Get last allocated variable.
+            template<typename T>
+            T& getLatestVar() const
+            {
+                return *reinterpret_cast<T*>(&m_mem[m_allocdBytes - sizeof(T)]);
+            }
+
+#if BOOST_COMP_GNUC
+#    pragma GCC diagnostic pop
+#endif
+            //! write initial meta data header
+            //!
+            //! Memory layout after the initialization.
+            //! |Header|....uninitialized Data ....
+            void init() const
+            {
+                // Initialize the data area with an empty meta data header.
+                m_allocdBytes = varChunkSize<MetaData>(0);
+                MetaData& ptr = getLatestVar<MetaData>();
+                ptr = MetaData{};
+            }
+
+            //! Memory chunk size required to store given variable.
+            //!
+            //! Calculate bytes required to store a type with a aligned starting offset.
+            //! Start offset to the origin of the user data chunk can be calculated with `result - sizeof(T)`.
+            //! The padding is always before the origin of the user data chunk and can be zero byte.
+            //!
+            //! \tparam T type should fit into the chunk
+            //! \param byteOffset Current byte offset.
+            //! \result Byte offset to the end of the user data chunk.
+            template<typename T>
+            std::uint32_t varChunkSize(uint32_t byteOffset) const
+            {
+                static_assert(
+                    core::vectorization::defaultAlignment >= alignof(T),
+                    "Unable to get block shared static memory for types with alignment higher than defaultAlignment!");
```

Comment:
I believe we should check against `TDataAlignBytes` here, which only happens to default to `core::vectorization::defaultAlignment`.

Reply:
I have not changed this code but this check is valid.
I think you assumed at the moment where you wrote this code that `m_mem` is default aligned https://github.com/alpaka-group/alpaka/blob/ed3c123b22684b9a342506d4d46a03a8ac6bd9bf/include/alpaka/block/shared/dyn/BlockSharedMemDynMember.hpp#L98-L102
Therefore it is possible to do the alignment based on an offset relative to `m_mem`.
It would be possible to remove the assert fully if we always compute the alignment based on `m_mem + byte_offset`.

Changing the assert to `>= TDataAlignBytes` would mean you can with OpenMP5 only handle aligned data up to `4` because of https://github.com/alpaka-group/alpaka/blob/ed3c123b22684b9a342506d4d46a03a8ac6bd9bf/include/alpaka/block/shared/st/BlockSharedMemStOmp5.hpp#L31

Reply:
Good point. I would actually chalk this up as a bug in the OMP5 ( [and OpenACC](https://github.com/alpaka-group/alpaka/blob/ed3c123b22684b9a342506d4d46a03a8ac6bd9bf/include/alpaka/ctx/block/CtxBlockOacc.hpp#L43) backends: I thought about 4-byte banks on GPUs and never came back to fix this. The intention of the template parameter was to set the alignment of the memory pool, which would limit the alignment of the variables because alignment was done relative to this address. If we can safely align the pointer address, we can indeed remove the assert which would make the template parameter obsolete, too.

... Actually, the template parameter `TDataAlignBytes` is already obsolete, because the alignment of the memory is hardcoded:
https://github.com/alpaka-group/alpaka/blob/develop/include/alpaka/block/shared/dyn/BlockSharedMemDynMember.hpp#L49

Please remove `TDataAlignBytes` and the assert.

Reply:
`TDataAlignBytes` is used to enforce a minimum alignment. Also, the static assert in the current version is correct, as you linked the dynamic shared memory object is aligned to `core::vectorization::defaultAlignment`. Therefore we can not guarantee any alignment larger than this.
To have a minimum alignment of 4byte for OpenMP which can be offloaded to GPUs sound valid, if this should be changed we should do it after this PR in a separate PR to not hide this change within this PR.

Reply:
Right. I missed the `std::max`. As a minimum alignment it should stay. I have to say, the name is a bad choice then. Could you change it to `TMinDataAlignBytes` in this PR @psychocoderHPC ?